src/evo/dmn_types.h

```diff
@@ -6,39 +6,50 @@
 #define BITCOIN_EVO_DMN_TYPES_H
 
 #include <amount.h>
+#include <serialize.h>
+
 #include <cassert>
 #include <string_view>
 
-class CDeterministicMNType
-{
-public:
-    uint8_t index;
-    int32_t voting_weight;
-    CAmount collat_amount;
-    std::string_view description;
+enum class MnType : uint16_t {
+    Regular = 0,
+    HighPerformance = 1,
 };
 
-namespace MnType {
-constexpr auto Regular = CDeterministicMNType{
-    .index = 0,
-    .voting_weight = 1,
-    .collat_amount = 1000 * COIN,
-    .description = "Regular",
-};
-constexpr auto HighPerformance = CDeterministicMNType{
-    .index = 1,
-    .voting_weight = 4,
-    .collat_amount = 4000 * COIN,
-    .description = "HighPerformance",
+template<> struct is_serializable_enum<MnType> : std::true_type {};
+
+struct CMnType
+{
+    const int32_t voting_weight;
+    const CAmount collat_amount;
+    const std::string_view description;
+
+    [[nodiscard]] static constexpr CMnType Regular() {
+        return CMnType{
+            .voting_weight = 1,
+            .collat_amount = 1000 * COIN,
+            .description = "Regular",
+        };
+    }
+    [[nodiscard]] static constexpr CMnType HighPerformance() {
+        return CMnType{
+            .voting_weight = 4,
+            .collat_amount = 4000 * COIN,
+            .description = "HighPerformance",
+        };
+    }
```

Comment:
It concerns me that we're potentially constructing this each time... What was wrong with the previous way of being inside a variable instead of a function?

Reply:
> It concerns me that we're potentially constructing this each time... 

good point 👍 we can do this to make sure it doesn't happen I think

```diff
     const CAmount collat_amount;
     const std::string_view description;
 
-    [[nodiscard]] static constexpr CMnType Regular() {
-        return CMnType{
+    [[nodiscard]] static constexpr CMnType Regular()
+    {
+        constexpr CMnType type = CMnType{
             .voting_weight = 1,
             .collat_amount = 1000 * COIN,
             .description = "Regular",
         };
+        return type;
     }
-    [[nodiscard]] static constexpr CMnType HighPerformance() {
-        return CMnType{
+    [[nodiscard]] static constexpr CMnType HighPerformance()
+    {
+        constexpr CMnType type = CMnType{
             .voting_weight = 4,
             .collat_amount = 4000 * COIN,
             .description = "HighPerformance",
         };
+        return type;
     }
```

Reply:
or maybe 1e0eeb81ab if you don't like having functions inside of a struct in general :)

Reply:
I tested and indeed without -O2 there's still a call of function: https://gcc.godbolt.org/z/WjfdTY9zc
Original Ody's version is also not fully optimized (still call GetMnType but do not construct an object from scratch, just return reference to pre-allocated flat memory): https://gcc.godbolt.org/z/Pnczeja3h


I updated branch! So far as we don't have `consteval` or any other way to force compiler to calculate it in compilation time, let it be.